warehouse/forklift/legacy.py

```diff
@@ -260,85 +260,70 @@ def _is_valid_dist_file(filename, filetype):
-        # Ensure the compression ratio is not absurd (decompression bomb)
-        compressed_size = os.stat(filename).st_size
-        with zipfile.ZipFile(filename) as zfp:
-            decompressed_size = sum(e.file_size for e in zfp.infolist())
-        if (
-            decompressed_size > COMPRESSION_RATIO_MIN_SIZE
-            and decompressed_size / compressed_size > COMPRESSION_RATIO_THRESHOLD
-        ):
-            sentry_sdk.capture_message(
-                f"File {filename} ({filetype}) exceeds compression ratio "
-                f"of {COMPRESSION_RATIO_THRESHOLD} "
-                f"({decompressed_size}/{compressed_size})"
-            )
+    if filename.endswith((".zip", ".whl")):
+        if not zipfile.is_zipfile(filename):
             return False
+        # Ensure that this is a valid zip file, and that it has a
+        # PKG-INFO or WHEEL file.
+        try:
+            with zipfile.ZipFile(filename) as zfp:
+                # Ensure that the compression ratio is not absurd (decompression bomb)
+                compressed_size = os.stat(filename).st_size
+                decompressed_size = sum(e.file_size for e in zfp.infolist())
+                if (
+                    decompressed_size > COMPRESSION_RATIO_MIN_SIZE
+                    and decompressed_size / compressed_size
+                    > COMPRESSION_RATIO_THRESHOLD
+                ):
+                    sentry_sdk.capture_message(
+                        f"File {filename} ({filetype}) exceeds compression ratio "
+                        f"of {COMPRESSION_RATIO_THRESHOLD} "
+                        f"({decompressed_size}/{compressed_size})"
+                    )
+                    return False
 
-        # Check that the compression type is valid
-        with zipfile.ZipFile(filename) as zfp:
-            for zinfo in zfp.infolist():
-                if zinfo.compress_type not in {
-                    zipfile.ZIP_STORED,
-                    zipfile.ZIP_DEFLATED,
-                }:
+                # Check that the compression type is valid
+                for zinfo in zfp.infolist():
+                    if zinfo.compress_type not in {
+                        zipfile.ZIP_STORED,
+                        zipfile.ZIP_DEFLATED,
+                    }:
+                        return False
+
+                # Check that the right files are present
+                for zipname in zfp.namelist():
+                    _, tail = os.path.split(zipname)
+                    if filename.endswith(".zip") and tail == "PKG-INFO":
+                        break
+                    if filename.endswith(".whl") and tail == "WHEEL":
+                        break
+                else:
                     return False
 
-    if filename.endswith(".tar.gz"):
-        # TODO: Ideally Ensure the compression ratio is not absurd
-        # (decompression bomb), like we do for wheel/zip above.
+        except zipfile.BadZipFile:  # pragma: no cover
```

Comment:
Flagging: I tried a few things to craft a zip file that passed `is_zipfile` but would raise this exception and couldn't. Keeping this here as a safeguard but skipping coverage checks for this line.

Reply:
Instead of the case of `is_zipfile()` check's `return False`, could you `raise zipfile.BadZipFile` instead, and have the `except` catch it?

Reply:
Why? I don't really think it's a great idea to manually raise non-generic exceptions like that, what would that get us?